### ISSUE TITLE
FakeCUDAKernel doesn't take "debug"

### ISSUE THREAD

**Issue by shaunc** [numba/numba#6615]
Using the CUDA simulator:

https://github.com/numba/numba/blob/ba0a6c8e7a7e33ad812bc644b29d2dfd8f3f4f0d/numba/cuda/decorators.py#L81-L82

and several other places pass debug=debug to `FakeCUDAKernel`, but it doesn't take `debug` option:

https://github.com/numba/numba/blob/ba0a6c8e7a7e33ad812bc644b29d2dfd8f3f4f0d/numba/cuda/simulator/kernel.py#L58

**Comment by esc** [comment]
@shaunc thank you for reporting this. Is there a specific use-case that breaks because of this?

**Comment by shaunc** [comment]
Hmm... well I'm now getting this whenever I use CUDASIM. However, I'm not sure exactly what is triggering it. I used CUDASIM about a month ago without problem, and both added a fair amount of code to my repo and also upgraded from 0.51.0 (...ish - I think)

**Comment by stuartarchibald** [comment]
Any chance of a reproducer please, just to make sure everyone is discussing the same thing. Thanks.

**Comment by shaunc** [comment]
I'd suggest you just need to extend your test suite for  to cover `config.ENABLE_CUDASIM`. It looks to me as if there isn't non-error any path through `@jit` wrapper in this case. (Presumably if coverage worked with numba, you would have caught it yourself -- that would be a feature request, but I think there is already an open issue for it.)

**Comment by stuartarchibald** [comment]
I suspect this could be involved.
```
$ python -c 'from numba import config; config.ENABLE_CUDASIM=1; from numba.cuda import jit; print(jit.__module__)'
numba.cuda.simulator.api

$ python -c 'from numba import config; config.ENABLE_CUDASIM=0; from numba.cuda import jit; print(jit.__module__)'
numba.cuda.decorators
```
The test suite already runs with `NUMBA_ENABLE_CUDASIM`:
https://github.com/numba/numba/blob/ba0a6c8e7a7e33ad812bc644b29d2dfd8f3f4f0d/buildscripts/incremental/test.sh#L82-L84
which makes me think that something else is quite possibly going on, either an untested path or an unusual set of circumstances.

A way to reproduce would be really useful.

**Comment by stuartarchibald** [comment]
@gmarkall is this branching (noted above) in use? https://github.com/numba/numba/blob/ba0a6c8e7a7e33ad812bc644b29d2dfd8f3f4f0d/numba/cuda/decorators.py#L79-L85

Is this in case someone bypasses the standard import mechanism which defers to an entirely separate decorator under the simulator?

**Comment by shaunc** [comment]
Hmm... so I use:
```
from numba import cuda as nc  # type: ignore
...
@nc.jit(
    "int8[:,:], float32[:], int8[:,:], int8[:,:,:], float32[:, :], float32[:], float32, float32"
)
```
Is this wrong? It would seem that perhaps cuda sim is relying unnecessarily on monkey patching?

Separately on the forum: https://numba.discourse.group/t/turn-on-cudasim-only-if-no-cuda/409
My larger issue is that I'd like to run my test suite including cuda code whether or not I'm on a computer that supports CUDA (I develop on a CUDA-less Mac, and by default test on push in gitlab workflow w/o access to cuda.) CUDA simulator seems to have been foreseen only for debugging, but IMO would also be very useful for testing. (Also separately, it is too permissive about what will actually run in CUDA, but that would be an incremental improvement.)

**Comment by seibert** [comment]
The main reason we are hesitant to encourage CUDASIM for testing is that it is not designed to be a detailed enough simulation of CUDA behavior to be reliable for testing.  The ability to insert unsupported code into CUDA functions when running the simulator is actually feature, since that enables things like using the Python debugger or other tools to trace through your function logic.  However, it will hide race conditions and other common CUDA mistakes that would be caught on a real device.

(Incidentally, this tension between debugging and testing eventually convinced NVIDIA to drop their equivalent to the CUDASIM feature in nvcc years ago.  Once they had cuda-gdb, the feature was considered more confusing than helpful.  If there were similar support for pdb connecting to CUDA devices, we might consider the same.)

**Comment by shaunc** [comment]
Thanks for the explanation. I have found it helpful -- for a while I was having to pay to spin up a CUDA staging environment in AWS just to test, and I found that testing on CUDA-simulator allowed me to "get onto the putting green" before I had to do this, even if it wasn't an exact simulation. Since then I bit the bullet and bought an in-house workstation to test using CUDA, but this isn't my normal workflow, and I am eager to find a way to continue to test in the cloud using CUDASIM (minimizing my AWS bills).

wrt to being able to support unsupported code in the simulator -- certainly that would be useful, but perhaps it could be dynamically toggled? e.g. --
```python
if tx == 1 and ty == 1:
    cuda.simulator.allow_unsupported()
    import pdb; pdb.set_trace()
    cuda.simulator.allow_unsupported(False)
```
Still, in the cases I actually hit, it was still quite useful as it is. I was able to debug errors in indexing, etc, and then when I went to real CUDA, then e.g. switching from `np.log2` to using `np.log( ) / np.log(2)` which slipped by in the simulator was simple. (Of course I understand that that might not always be the case.)

---
For the current bug, I guess this isn't relevant though. It would seem that perhaps by renaming on import I'm foiling the monkey patching? I guess use for testing would be just one reason not to use monkey patching, but there are others as well. :)
